File: 1380-number-of-closed-islands/number-of-closed-islands.cpp
class Solution {
public:
    int rows;
    int cols;

    bool dfs(int row, int col, vector<vector<int>>& grid) {
        if (row < 0 || row >= rows || col < 0 || col >= cols) {
            return false;
        }
        if (grid[row][col] == 1) {
            return true;
        }

        grid[row][col] = 1;

        bool down = dfs(row + 1, col, grid);
        bool up = dfs(row - 1, col, grid);
        bool right = dfs(row, col + 1, grid);
        bool left = dfs(row, col - 1, grid);

        return down && up && right && left;
    }

    int closedIsland(vector<vector<int>>& grid) {
        rows = grid.size();
        cols = grid[0].size();

        int closed = 0;
        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < cols; col++) {
                if (grid[row][col] == 0) {
                    if (dfs(row, col, grid)) {
                        closed++;
                    }
                }
            }
        }
        return closed;
    }
};

File: 1380-number-of-closed-islands/number-of-closed-islands.java
class Solution {
    public int rows, cols;

    public boolean DFS(int i, int j, int[][] grid, boolean[][] visited) {
        if(i < 0 || i >= rows || j < 0 || j >= cols) 
            return false;

        if(grid[i][j] == 1 || visited[i][j]) 
            return true;

        visited[i][j] = true;

        boolean DOWN = DFS(i + 1, j, grid, visited);
        boolean UP = DFS(i - 1, j, grid, visited);
        boolean RIGHT = DFS(i, j + 1, grid, visited);
        boolean LEFT = DFS(i, j - 1, grid, visited);

        return DOWN && UP && RIGHT && LEFT;
    }

    public int closedIsland(int[][] grid) {
        this.rows = grid.length;
        this.cols = grid[0].length;
        int isLandCount = 0;
        boolean[][] visited = new boolean[rows][cols];

        for(int i = 0; i < rows; i++) {
            for(int j = 0; j < cols; j++) {
                if(grid[i][j] == 0 && visited[i][j] == false) {
                    if(DFS(i, j, grid, visited)) {
                        isLandCount++;
                    }
                }
            }
        }
        return isLandCount;
    }
}